source/lac/sparse_direct.cc

```diff
@@ -1359,14 +1370,47 @@ SparseDirectMUMPS::vmult(VectorType &dst, const VectorType &src) const
       // Copy solution into the given vector
       // For MUMPS with centralized solution (icntl[20]=0), the solution is only
       // on the root process (0) and needs to be distributed to all processes
-      if (Utilities::MPI::this_mpi_process(mpi_communicator) == 0)
+
+      // Get locally owned range for this process
+      const IndexSet &locally_owned = dst.locally_owned_elements();
+      const size_type local_size    = locally_owned.n_elements();
+
+      std::vector<double> local_values(local_size);
+
+      const unsigned int my_rank =
+        Utilities::MPI::this_mpi_process(mpi_communicator);
+
+      int my_size_int  = local_size;
+      int my_start_int = locally_owned.nth_index_in_set(0);
+
+      std::vector<int> sizes =
+        Utilities::MPI::gather(mpi_communicator, my_size_int, 0);
+      std::vector<int> displs =
+        Utilities::MPI::gather(mpi_communicator, my_start_int, 0);
```

Comment:
You can make these variables `const`. I think this should fix the types:
```suggestion
      const std::vector<size_type> sizes =
        Utilities::MPI::gather(mpi_communicator, local_size, 0);
      const std::vector<types::global_dof_index> displs =
        Utilities::MPI::gather(mpi_communicator, locally_owned.nth_index_in_set(0);, 0);
```